feat(intrinsic_camera_calibrator): multiple camera models & ceres integration

amadeuszsz opened the pull request:
## Description

<!-- Write a brief description of this PR. -->

## Related links

<!-- Write the links related to this PR. -->

## Tests performed

<!-- Describe how you have tested this PR. -->

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

amadeuszsz committed:
feat(intrinsic_camera_calibrator): prototype for multiple camera models & ceres integration

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

pre-commit-ci[bot] committed:
ci(pre-commit): autofix

amadeuszsz committed:
fix(intrinsic_camera_calibrator): inverse flag logic

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

pre-commit-ci[bot] committed:
ci(pre-commit): autofix

amadeuszsz committed:
feat(intrinsic_camera_calibrator): handle camera model only in calibration settings

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): use desired camera model

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): use limited num of frames in init calibration for Ceres

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): suppress Ceres factorization warnings

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): update parameter

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): remove redundant parameter

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): remove unused variable

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): typos, cSpell

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): add coefficients regularization in Ceres

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(ceres_intrinsic_camera_calibrator): remove unused variables

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented on calibrators/intrinsic_camera_calibrator/ceres_intrinsic_camera_calibrator/include/ceres_intrinsic_camera_calibrator/coefficients_residual.hpp:
```diff
@@ -0,0 +1,140 @@
+    const T & k6 =
+      rational_distortion_coeffs_ > 2 ? camera_intrinsics[distortion_index++] : null_value;
+
+    residuals[0] = num_samples_factor_ * regularization_weight_ * pow(k1, 2);
```
@amadeuszsz 
Please have a look at:
http://ceres-solver.org/nnls_tutorial.html#hello-world

I am very sure the pow is not needed here

knzo25 replied:
Also, the `num_samples` and `regularization_weight` are being squared.
As far as I know, the way to apply weights is something like:
https://github.com/tier4/CalibrationTools/blob/dc1bd3a677e10a89e4324135083d0187c18b2abb/calibrators/tag_based_sfm_calibrator/src/ceres/calibration_problem.cpp#L690

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/2e26b158b248757abd99ce41ca6f9a013c77a6fa

knzo25 commented on calibrators/intrinsic_camera_calibrator/ceres_intrinsic_camera_calibrator/include/ceres_intrinsic_camera_calibrator/coefficients_residual.hpp:
```diff
@@ -0,0 +1,140 @@
+#ifndef CERES_INTRINSIC_CAMERA_CALIBRATOR__COEFFICIENTS_RESIDUAL_HPP_
+#define CERES_INTRINSIC_CAMERA_CALIBRATOR__COEFFICIENTS_RESIDUAL_HPP_
+
+#include <Eigen/Core>
+#include <opencv2/core.hpp>
```
Needed?

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/055aa6c5f43e13789aff6017c0ad8c444c66de7c

amadeuszsz committed:
docs(ceres_intrinsic_camera_calibrator): add missing docs

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented on calibrators/intrinsic_camera_calibrator/ceres_intrinsic_camera_calibrator/include/ceres_intrinsic_camera_calibrator/coefficients_residual.hpp:
```diff
@@ -0,0 +1,140 @@
+#include <ceres/autodiff_cost_function.h>
+#include <ceres/ceres.h>
+
+struct CoefficientsResidual
```
nit: the struct name does not reflect the fact that we are referring to the coefficients of the distortion model. DistortionResidual or DistortionCoefficientsResidual would be better imo

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/ea9978619505086ffb1610d81360bb294318134e

knzo25 commented on calibrators/intrinsic_camera_calibrator/ceres_intrinsic_camera_calibrator/src/ceres_intrinsic_camera_calibrator_py.cpp:
```diff
@@ -182,14 +184,16 @@ PYBIND11_MODULE(ceres_intrinsic_camera_calibrator_py, m)
             num_radial_coeffs (int): The number of radial distortion coefficients used during calibration
             num_rational_coeffs (int): The number of rational distortion coefficients used during calibration
             use_tangential_distortion (bool): Whether we should use tangential distortion during calibration
+            regularization_weight (float): The regularization weight for distortion coefficients
```
was it not set to double?

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/ca7b41c45b9f5dc21ce09fc17fa98bb0643154b6

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_models/opencv_camera_model.py:
```diff
@@ -0,0 +1,85 @@
+    ):
+        """Update parameters."""
+        for idx, k in enumerate([cv2.CALIB_FIX_K1, cv2.CALIB_FIX_K2, cv2.CALIB_FIX_K3]):
+            self.flags = toggle_flag(self.flags, k, not (idx < radial_distortion_coefficients))
```
these logical statements could be simplified

amadeuszsz replied:
Did you mean `not`? It was done for readability purposes - parameter not set via GUI means flag set for OpenCV.

knzo25 replied:
Ok, I guess logic statements' readability changes person to person xd

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_models/ceres_camera_model.py:
```diff
@@ -0,0 +1,118 @@
+from intrinsic_camera_calibrator.camera_models.opencv_camera_model import OpenCVCameraModel
+import numpy as np
+
+os.environ["GLOG_minloglevel"] = "2"  # cSpell:ignore minloglevel
```
Can you add this as an optional step linked to the verbosity of the calibrator?

amadeuszsz replied:
We do not handle verbosity via arguments. I can add `glog_level` parameter to config file, but setting environment variable still has to be done before importing Ceres library bindings. This force us to place `import` e.g. in class constructor, which is far away from being *pythonic*. Before I will do it, let me know how you see it because there is many places where we can do it, none of them are OK for me.
The best solution could be setting verbosity directly in our Ceres library, however glog ignores it for me and that's why I suppress optimization warnings by setting environment variable.

amadeuszsz replied:
Ok, I found this https://github.com/tier4/CalibrationTools/pull/208/commits/34679cb30708feb3e26624c318b5db30ba8f0809
For me this is acceptable

knzo25 commented:
@amadeuszsz 
remember to link the related tickets and data needed to reproduce (so I can use the same data and independent one if possible)

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/parameter.py:
```diff
@@ -89,3 +89,14 @@ def parameters(self) -> dict:
+    def get_parameters_values(self) -> dict:
+        """Return the values of the Parameter objects from this class as a dictionary."""
+        with self.lock:
+            p_dict = {}
```
nit: this code is almost the same as the previous one.
Code can be reused and then extended with a comprehension:

e.g., `return {k: v.value for k,v in self.parameters()}`

knzo25 replied:
(it could be argued that the original implementation is also suited for a comprehension)

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/184701ce4a34037ce28b397f4e611692f11f35d2

knzo25 replied:
@amadeuszsz 
`parameters` already uses a lock so the one in `get_parameters_values` should not be needed.
We are using RLock so while it may not be needed, I think it is better to remove the second `with self.lock`

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/237330de94fb6dd773c9511270f002aac733b6ff

knzo25 commented:
@amadeuszsz 
Regarding:
`intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/board_detections/board_detection.py`,

The several statistics related to the pose, angles, etc that require a camera model, should use "the best camera model available", which is not the single shot. Single shot works well for these only in the case that the camera is not distorted.

Single shot should be used to detect outliers based on the reprojection error. E.g., if a model can not fit even one sample, it is highly probable that that sample is an outlier

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/calibrators/ceres_calibrator.py:
```diff
@@ -18,19 +18,52 @@
+        self.use_tangential_distortion = Parameter(
+            bool, value=True, min_value=False, max_value=True
+        )
+        self.pre_calibration_num_samples = Parameter(int, value=6, min_value=0, max_value=20)
```
This number should be as high as possible taking into consideration that opencv does not scale well in terms of computational complexity (e.g., we limit it to 80 samples by default. when sergio enables a few decomposition options it will increase to a few hundred)

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/1ad49f1b1f6a38b01b3cdf7e96508469006fafee

knzo25 replied:
Sorry for being insistent, but can you leave it as a higher number? (in the experiments I remember doing with over 40 is was way more stable. How was it for you?)

amadeuszsz replied:
Confirmed, addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/e5a1de6b5896d904da39babd9618356680951e8e

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_models/ceres_camera_model.py:
```diff
@@ -0,0 +1,118 @@
+        width: Optional[int] = None,
+    ):
+        super().__init__(k, d, height, width)
+        self.radial_distortion_coefficients = 3
+        self.rational_distortion_coefficients = 3
+        self.use_tangential_distortion = True
+        self.pre_calibration_num_samples = 6
+        self.regularization_weight = 0.001
```
having these values defined here makes it difficult to understand. Everything should be loaded from the `Parameter` class

amadeuszsz replied:
This class does not inherit from Qt, I think it's not necessary since Calibrator does. It is independent from Qt Parameter. We can remove default values for better readability - calibrators contain default values.
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/47d0c16ee070724e64fe386f43b47aa45eb3a750

knzo25 replied:
nit:

Regarding typing:
`int_value: int = None`
should be
`int_value: Optional[int] = None`

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/b6168317810cfd8abf813aa2d7598b9940111530

knzo25 reviewed (changes requested):
@amadeuszsz 
Confirmed basic functionalities, but got enough comments that I feel I would like to see them addressed or dicussed before further reviews

amadeuszsz committed:
fix(ceres_intrinsic_camera_calibrator): remove unused headers

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
refactor(ceres_intrinsic_camera_calibrator): change struct name

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
refactor(ceres_intrinsic_camera_calibrator): remove squaring

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
docs(ceres_intrinsic_camera_calibrator): typo

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
refactor(intrinsic_camera_calibrator): use of comprehension

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): update parameter max value

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
refactor(ceres_intrinsic_camera_calibrator): apply regularizaion weight for distortion coefficients via loss function

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): assign default values only once

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
style(intrinsic_camera_calibrator): pre-commit

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz commented:
> remember to link the related tickets and data needed to reproduce (so I can use the same data and independent one if possible)

@knzo25 
The most important / unique data is for regularization, which isolates unstable coefficients issue. It was attached in this ticket [INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-8307)
Also updated link for all the data used in experiments in this ticket [INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-8305)

amadeuszsz committed:
feat(intrinsic_camera_calibrator): logging severity

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): cspell

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): unecessary lock

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented:
@amadeuszsz 
Regarding my previous comment:

> @amadeuszsz
Regarding:
intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/board_detections/board_detection.py,

> The several statistics related to the pose, angles, etc that require a camera model, should use "the best camera model available", which is not the single shot. Single shot works well for these only in the case that the camera is not distorted.

> Single shot should be used to detect outliers based on the reprojection error. E.g., if a model can not fit even one sample, it is highly probable that that sample is an outlier


This may be too big for this PR since also affects some of sergio's work. In the meantime, can you turn all data collection filters that use 3D information ? (pose, 3d rotation, etc). As explained before, using single shot models to estimate these values with highly distorted cameras will end up with trashy values

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_models/ceres_camera_model.py:
```diff
@@ -0,0 +1,117 @@
+            use_tangential_distortion=self.use_tangential_distortion,
+        )
+
+        # Consider only part of data (distributed uniformly)
```
Is this assumption valid? (I do not remember implementing shuffling :thinking: )

amadeuszsz replied:
For initial calibration I assume it is better to take uniform distributed (from collecting sequence point of view) samples, e.g. If we have 80 samples and 40 samples limit for initial calibration, we take 1, 3, 5... instead of 1-40. We could take samples base on pose distribution as well, but I didn't find it necessary.
In general, I would prefer to remove this limit in favor of timeout parameter, but OpenCV API does not consider this. Separate thread for init calibration task would not work - breaking it will not result with recent intrinsic parameters. Maybe calibrating frame by frame here is an answer, but I doubt the performance.

amadeuszsz committed:
feat(intrinsic_camera_calibrator): update default parameter value

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/board_detections/board_detection.py:
```diff
@@ -179,11 +165,8 @@ def get_pose(self, model: Optional[CameraModel] = None) -> Tuple[np.array, np.ar
         return self._cached_pose
 
-    def get_flattened_3d_points(self, model: Optional[CameraModel] = None) -> np.array:
+    def get_flattened_3d_points(self, model: CameraModel) -> np.array:
```
The following lines are a bug from before this PR, but can you fix them here?

```
if model == self._cached_camera_model and self._cached_flattened_3d_points is not None:
            return self._cached_pose
```

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/d966e0376241ef319b1b72fc814da0f36be0eeba

amadeuszsz committed:
style(intrinsic_camera_calibrator): var type

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/board_detections/board_detection.py:
```diff
@@ -108,33 +107,26 @@ def get_center_2d(self) -> np.array:
-            and model == self._cached_camera_model
-            and self._cached_reprojection_errors is not None
-        ):
+        if self._cached_camera_model is None:
```
Even if we do not apply the full behavior regarding the 3d statistics that require a non-single shot calibration, the following should still be implemented in this PR:

Have:
`get_reprojection_errors`: uses an existing model
and
`get_single_shot_reprojection_errors`: forces calibration using a single sample. It still requires an input model to know which kind of model it should use (parameters ,etc)

Alternatively, use only `get_reprojection_errors` and implement the required behavior in the caller.

Currently, there are two kind of uses for `self._cached_camera_model`. Maybe the should be split.

knzo25 replied:
In this method, what happens if:
`self._cached_camera_model is not None` and `model != self._cached_camera_model` ?

amadeuszsz replied:
> get_reprojection_errors: uses an existing model

Using existing model seems working here after changes in camera model (using last camera matrix & coefficients), but after triggering calibration we obtain bad results. There are some assumptions in `_calibrate` within calibrator and mentioned change in camera model is breaking change.  I would need to reconsider design here again.

Instead of refactor twice, I would prefer to address this issue with [this](https://github.com/tier4/CalibrationTools/pull/208#issuecomment-2487368546) at once in separate PR if it is not a problem.

> In this method, what happens if:
> self._cached_camera_model is not None and model != self._cached_camera_model ?

Model has been changed, so we want to update our reprojection error. Isn't it expected behavior? We can change it.

amadeuszsz committed:
fix(intrinsic_camera_calibrator): return correct value

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/calibrators/ceres_calibrator.py:
```diff
@@ -18,19 +18,52 @@
+            bool, value=True, min_value=False, max_value=True
+        )
+        self.pre_calibration_num_samples = Parameter(int, value=6, min_value=1, max_value=100)
+        self.regularization_weight = Parameter(float, value=0.001, min_value=0.0, max_value=1.0)
```
Leaving a placeholder comment to remind ourselves to leave the default values and the max value after seeing the experiments related to this

knzo25 replied:
Marking this as resolved since this value seemed like a very good one (internal evidence)

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/calibrators/opencv_calibrator.py:
```diff
@@ -31,42 +33,41 @@ class OpenCVCalibrator(Calibrator):
         width = detections[0].get_image_width()
 
-        camera_model = CameraModel()
+        camera_model = make_camera_model(camera_model_type=CameraModelEnum.OPENCV)
```
This seems to be a problem from before, but can you check if all `.value` accesses  are done in a locked context?
Applies to the calibration related methods

amadeuszsz replied:
I just checked it and it seems `.value` for all parameters from GUI are reflected here.

knzo25 replied:
That is not what I meant.
The parameters are accessed both in the GUI threads and the calibration ones. This raises potential dataraces.
So far I had been making sure (or at least tried) that all accesses to `.value ` were done in a `with self.lock:` context.

I just checked, and this one seems is not. Can you address these cases?
Alternatively, the lock part of ParameterizedClass could be made to be applied to the Parameter class

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/73eb77bd6fbfb12e67a5707eeddc1082e8fe91d1

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_calibrator.py:
```diff
@@ -648,7 +655,8 @@ def start(
         self.board_type = board_type
         self.board_parameters = board_parameters
-        self.current_camera_model = initial_intrinsics
+        if initial_intrinsics is not None:
```
I am lost. Why is this needed?

amadeuszsz replied:
Ahh, it's not. Probably remains from first commit which consist of different camera model design.
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/50d013f5ee80717c808cdc6ddbe7e95256eebbfc

knzo25 commented on calibrators/intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/camera_models/ceres_camera_model.py:
```diff
@@ -0,0 +1,117 @@
+        """Calibrate Ceres camera model."""
+        camera_model = self.init_calibrate(object_points_list, image_points_list)
+
+        rms_error, camera_matrix, dist_coeffs, rvecs, tvecs = calibrate(
```
CI/CD may complain in the future about unused variables, so could you `_` them ?

amadeuszsz replied:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/76f361388c16eb5663527f167a48557b47c1685e

knzo25 commented:
@amadeuszsz 
Regarding how the distortion coefficients are treated inside the ceres calibrator.
Kind of got the impression there would be errors if `radial coefficients < 3` and `rational cofficients > 0`.
Could you check? If seems to much a trouble, we could add a restriction that makes `radial coefficients == 3` if we want to use rational coefficients (personally, I would prefer this :sweat_smile: )

amadeuszsz committed:
fix(intrinsic_camera_calibrator): revert unnecessary change

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
fix(intrinsic_camera_calibrator): mark unused variables

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): update default value

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz commented:
@knzo25 

> Regarding how the distortion coefficients are treated inside the ceres calibrator. Kind of got the impression there would be errors if `radial coefficients < 3` and `rational cofficients > 0`. Could you check? If seems to much a trouble, we could add a restriction that makes `radial coefficients == 3` if we want to use rational coefficients (personally, I would prefer this 😅 )

Why? I checked and it works for any combination. Also I can see how reprojection error changes so I believe it is due to camera model configuration.

> Regarding:
> > intrinsic_camera_calibrator/intrinsic_camera_calibrator/intrinsic_camera_calibrator/board_detections/board_detection.py,
> >
> > The several statistics related to the pose, angles, etc that require a camera model, should use "the best camera model available", which is not the single shot. Single shot works well for these only in the case that the camera is not distorted.
>> 
>> Single shot should be used to detect outliers based on the reprojection error. E.g., if a model can not fit even one sample, it is highly probable that that sample is an outlier
> 
> This may be too big for this PR since also affects some of sergio's work. In the meantime, can you turn all data collection filters that use 3D information ? (pose, 3d rotation, etc). As explained before, using single shot models to estimate these values with highly distorted cameras will end up with trashy values

Yup, use recent best model brings me some issues and we would need to redesign some part of the code to assign camera model configuration to board detections. Your temporary workaround is addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/494fcb46736c3c6c5d7062ac420678a76e18eb77

knzo25 commented:
Following occam's razor principle, I would like to leave the default distortion model fixed at radial coefficients=2 with no rational coefficients.

For ceres, the max calibration samples, currently capped at 80, should be increased to a large number, 500 or 1000 for example

Final comment. Not sure it this happened now or before (at least when I first implemented this, it did not happen), but in the current code, the loggings are not really working. Can you make sure that the ones in the calibration routine are flushed (shown in the console) as soon as they are executed?

knzo25 reviewed (changes requested):
There is one comment from before left, plus some minor changes related to parameters and logging :pray:

amadeuszsz committed:
fix(intrinsic_camera_calibrator): lock thread while read parameters

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): update default parameters

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): apply logging severity for Python logging

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz committed:
feat(intrinsic_camera_calibrator): override default number of calibration samples for Ceres calibrator

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz commented:
@knzo25 

> Following occam's razor principle, I would like to leave the default distortion model fixed at radial coefficients=2 with no rational coefficients.

Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/8d990f011f2b20de69de4e41d1e76b24c67bf7f2

> For ceres, the max calibration samples, currently capped at 80, should be increased to a large number, 500 or 1000 for example

Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/ef56973de8a4ba26221ed45ec697963445ab4532

> Final comment. Not sure it this happened now or before (at least when I first implemented this, it did not happen), but in the current code, the loggings are not really working. Can you make sure that the ones in the calibration routine are flushed (shown in the console) as soon as they are executed?

Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/4e3a48d88b47b017c617c1942c46ae8cf2745fd3
Now we use same logging severity as it was set for glog (Ceres c++ bindings). Default is `2` (error) and can be easily controlled via CLI. Please, let me know if you would like to change this default value.

amadeuszsz committed:
style(intrinsic_camera_calibrator): cspell

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

knzo25 commented:
@amadeuszsz 
Regarding https://github.com/tier4/CalibrationTools/commit/ef56973de8a4ba26221ed45ec697963445ab4532
I mean that number as the default parameter of what goes after the entropy subsampling, In my test even if,say, 200 samples passed the pre-rejection flter, only 80 were used for the final ceres calibration.

regarding logging, only `cvUndistortPointsInternal` is displayed whereas all the other relevant logging does not appear

amadeuszsz commented:
@knzo25 

> Regarding [ef56973](https://github.com/tier4/CalibrationTools/commit/ef56973de8a4ba26221ed45ec697963445ab4532) I mean that number as the default parameter of what goes after the entropy subsampling, In my test even if,say, 200 samples passed the pre-rejection flter, only 80 were used for the final ceres calibration.

Could you show me number of post-rejection inliers? The value of `80` was occurring only in addressed commit and after update I can see more calibration samples used for calibration with default parameter (`500`).

![Screenshot from 2024-11-25 22-12-55](https://github.com/user-attachments/assets/851194ec-6e76-4988-9a33-33a446884ad3)

> regarding logging, only `cvUndistortPointsInternal` is displayed whereas all the other relevant logging does not appear

I can see every single log using appropriate severity level, e.g.:
```bash
[camera_calibrator-1] INFO:root:Iteration 93: inliers: 123 | mean rms: 0.62 | min rms: 0.07 | max rms: 7.94                                                                                                                                                                        
```
Did you run launch file with custom severity?
```bash
ros2 launch intrinsic_camera_calibrator calibrator.launch.xml severity:=0
```

knzo25 commented:
The error with `cvUndistortPointsInternal` was due to a custom opencv I had installed. Apologies for that one
Regarding the `max_calibration_samples`, I misunderstood its current meaning (now there are 3 limits), but can you please delete the `self.max_calibration_samples.value = 500`?

If you can also set `<arg name="severity" default="0"/>`  we can merge this PR!

Note: I was getting very bad detection results for the chess board, but it seems it is unrelated to the PR

amadeuszsz committed:
feat(intrinsic_camera_calibrator): update params

Signed-off-by: amadeuszsz <amadeusz.szymko.2@tier4.jp>

amadeuszsz commented:
> but can you please delete the `self.max_calibration_samples.value = 500`?
> 
> If you can also set `<arg name="severity" default="0"/>` we can merge this PR!

Done

knzo25 reviewed (approved):
LGTM !
Thanks for the cleanest PR I have seen in a while ahaha
